Added new workshop - Interactive Maps (#1342)
hackclub/hackclub: workshops/interactive_maps/README.md

| Name      | Description |Author|
| :----: | :----: | :----: |

> | Interactive Maps Using JS | Learn JS and HTML while making a fun interactive map | @LonaDotExe |

Reviewer: incorrect formatting, pls check other workshops to see correct method. it should be yaml

- Final code - https://repl.it/@JeswinSunsi/Interactive-Maps#index.html

![gif demo](https://cloud-fmga8dmlt.vercel.app/gif_one.gif)

## 1 - Prerequisites

- Basic understanding of functions
- Amateur knowledge of HTML and CSS

In case you don't know either of those, don't worry. We'll go through each line of code in depth.

## 2 - Tools and Resources

- For this tutorial, we'll be using repl.it, an online browser-based IDE
- For editing the photos, we can use Pixlr.com or any other full fledged photo editor
- I recommend going through https://www.w3schools.com/js/js_functions.asp for further clarification with the JS part

## 3 - Designing the image

Any suitable image will work, but I have chosen an animated isometric map. The image was of the JPG format, but I removed the background to get a png image. Then from that one image, I selected specific areas and created 5 different images, each with a different area highlighted in red.
For a visual understanding, here is the original image followed by an edited one.

![image 1](https://cloud-fyde5ir0a.vercel.app/image.png)
![image 2](https://cloud-jmd1dy3k8.vercel.app/image.png)

> Basically, you'll have to take an image, and seperately highlight/color parts of it, and save it as different images. Here are the image resources I'm using:

Reviewer 2: ```suggestion
Basically, you'll have to take an image, and separately highlight/color parts of it, and save it as different images. Here are the image resources I'm using:
```

    }   
    
    else if (location == 'beach'){
        map.src = 'https://cloud-3bvmpy2n1.vercel.app/orgbeach.png';
    }
    else if (location == 'normal'){
        map.src = 'https://cloud-coh6o3mhy.vercel.app/orgbg.png';
    }
}
```
This snippet declares a variable named map, which corresponds to the map element in our html file. Then it creates a function named changeImage() with the parameter 'location'. When we call this function from our html file with specific values like 'ocean' or 'residential', (These values passed into parameters are called arguments.) they call on the actual function from the JS file. Inside the function, is an IF -ELSE IF statement. It does exactly what it means. If the location argument is 'beach', set the map's src attribute to the beach highlighted image, and so on. The syntax is
```js
if (condition) {
    Do something;
}
else if (another condition) {
    Do something else;
}
else {
    Do something totally different;
}
```
### Conclusion
And just like that, you have gone through the basics of HTML/CSS and a bit of JS to create your own interactive map. Now it's up to you to add in a bit of your sparkle and tweak it.

> ##### Hacking it

Reviewer: need three examples with demo links + source code

Reviewer: still don't see anything here

Author: Oh you mean three example website? Oops, I thought u meant the If statements. Workin on it

Author: Heyyo @sampoder, I've pushed another commit. Is this what you wanted?

Reviewer: yep, right style just the original demo doesn't count,

thanks

---
name: 'Interactive Maps'
description: 'Learn JS and HTML while making a fun interactive map'

> author: '@LonaDotExe'

Reviewer 3: The author name should be same as your github username. And in that case it is JeswinSunsi right?

---

# Making an interactive map using Javascript 

> By the end of this tutorial, you will learn to use functions and IF-ELSE statements in nothing but Vanilla JS to create some really good-looking interactive maps. This workshop will also teach you the basics of HTML and CSS. As a bonus, read till the end for a fun programming fact.

Reviewer 3: IF-ELSE --> if-else

- Final code - https://repl.it/@JeswinSunsi/Interactive-Maps#index.html

![gif demo](https://cloud-fmga8dmlt.vercel.app/gif_one.gif)
### 1 - Prerequisites
- Basic understanding of functions 
- Amateur knowledge of HTML and CSS

In case you don't know either of those, don't worry. We'll go through each line of code in depth.

### 2 - Tools and Resources
- For this tutorial, we'll be using repl.it, an online browser-based IDE
- For editing the photos, we can use Pixlr.com or any other full fledged photo editor
- I recommend going through https://www.w3schools.com/js/js_functions.asp for further clarification with the JS part

### 3 - Designing the image 
Any suitable image will work, but I have chosen an animated isometric map. The image was of the JPG format, but I removed the background to get a png image. Then from that one image, I selected specific areas and created 5 different images, each with a different area highlighted in red.
For a visual understanding, here is the original image followed by an edited one.

![image 1](https://cloud-fyde5ir0a.vercel.app/image.png)
![image 2](https://cloud-jmd1dy3k8.vercel.app/image.png)

### 4 - Coding it up
##### The boilerplate

> Head over to https://repl.it. Creating an account is of personal preference, be warned that you may lose your work if you don't have an account. Click on start coding and search for HTML, CSS, JS. Once you create a repl, you'll end up at this page.

Reviewer 3: you can tell the people to use this link https://repl.it/languages/html
instead of telling them to search for it, It will be easy for them that way!

![html boilerplate](https://cloud-5egibvgdv.vercel.app/image.png)

> As you can see, repl has already created the boilerplate code for your project. Let's take a quick look. 

Reviewer 3: If u want to add a line break here, use two spaces at the end of the line otherwise it won't get on the next line

> ```<!DOCTYPE html>```  is a declaration line that means we'll be coding in HTML 5.0. Following up are the ```<html>``` and ```<head>``` tags. Our html markup will be contained inside the two ```<html>``` tags. The ```<head>``` tag is where we will put some of our specifications which we normally don't want the end-user to see. The first ```<meta>``` tag tells the browser which character set to use. Here, it is set to UTF-8. The next ```<meta>``` tag specifies the page's dimensions and scaling. There is no problem in not understanding how these tags work, as they aren't very important. ```<title>``` is something that you should probably change. It is the text that is shown on the browser tab. Here it is set to repl.it. Coming next up in the ```<head>``` tag is ```<link>```. It tells the browser to load the CSS styling from the style.css file. It also informs the browser that it a css text file.

Reviewer 3: For all the code related terms, please use single backticks instead of triple backticks. Triple backticks are used for code blocks. If it is a single word or so, make sure to use a single backtick

> ```<!DOCTYPE html>```  is a declaration line that means we'll be coding in HTML 5.0. Following up are the ```<html>``` and ```<head>``` tags. Our html markup will be contained inside the two ```<html>``` tags. The ```<head>``` tag is where we will put some of our specifications which we normally don't want the end-user to see. The first ```<meta>``` tag tells the browser which character set to use. Here, it is set to UTF-8. The next ```<meta>``` tag specifies the page's dimensions and scaling. There is no problem in not understanding how these tags work, as they aren't very important. ```<title>``` is something that you should probably change. It is the text that is shown on the browser tab. Here it is set to repl.it. Coming next up in the ```<head>``` tag is ```<link>```. It tells the browser to load the CSS styling from the style.css file. It also informs the browser that it a css text file.

Reviewer 3: Maybe also split them into different paras

> Here the ```<head>``` tag ends, giving way to the body tag.  ```<body>``` is where we write the markup that the viewer will see. Repl.it has already created a ```<script>``` tag here that imports the JS code. Now that we have finished with understanding the boilerplate, let's start writing some of our own code.

Reviewer 3: If it is a beginner doing your workshop, he/she/they won't know where the head tag is closing. it will be good if you specify the closing tags. eg. </head>

##### The HTML Markup
As we can see from the final website gif, our site will have two parts, the first part which houses the buttons, and the second part, which houses the actual map. Let's get on with writing the code for the buttons. 

We can use HTML default buttons, but they look like they should've been abandoned along with Windows 1999, so let's spice things up a little and design some buttons of our own. Above the ```<script>``` tag and below the opening ```<body>``` tag, add this.
```html

>     <div class='first' style="margin-left:auto;margin-right:auto;">

Reviewer 3: there is no need for indentation here since there is no parent element in this particular code block.

      <a onmouseover="changeImage('ocean')" class="button pulse">Ocean</a>
      <a onmouseover="changeImage('beach')" class="button pulse">Beach</a>
      <a onmouseover="changeImage('roadway')" class="button pulse">Roadway</a>
      <a onmouseover="changeImage('vendors')" class="button pulse">Vendors</a>
      <a onmouseover="changeImage('residential')" class="button pulse">Residential</a>
    </div>
```

> Going through the code, we find a ```<div>```. Divs are a way of grouping multiple obects into one. Divs also have various styling properties. This div has two attributes, class and style. Class is just a way to apply similar styles to multiple objects in CSS. The second attribute is style. It's a way of directly styling an element inside HTML. Margin-left and Margin-left sets the margins of the div. Here, to auto. Read more about margins at https://www.w3schools.com/css/css_margin.asp. Inside the Div, we have 5 ```<a>``` tags. We will tweak these tags to create the buttons, although some other tags can be used for this too. All these buttons share the same classes -> button & pulse. the other attribute, onmouseover, does exactly what it means. It calls a javascript function when the mouse is over that particular element. Right now if you run the markup, it'd be a total mess. So let's add the rest of it. Immediatly under the closing ```</div>``` tag, add 

Reviewer 3: ```suggestion
Going through the code, we find a `<div>` tag. Divs are a way of grouping multiple obects into one. Divs also have various styling properties. This div tag has two attributes, class and style. Class is just a way to apply similar styles to multiple objects in CSS. The second attribute is style. It's a way of directly styling an element inside HTML. Margin-left and Margin-left sets the margins of the div. Here, to auto. 

Read more about margins at https://www.w3schools.com/css/css_margin.asp. 

Inside the Div, we have 5 ```<a>``` tags. We will tweak these tags to create the buttons, although some other tags can be used for this too. All these buttons share the same classes -> button & pulse. the other attribute, onmouseover, does exactly what it means. It calls a javascript function when the mouse is over that particular element. Right now if you run the markup, it'd be a total mess. So let's add the rest of it. Immediatly under the closing ```</div>``` tag, add 
```

  }
  33% {
    transform: scale(1);
    opacity: 1;
  }
  100% {
    transform: scale(3);
    opacity: 0;
  }
}
.button {
  display: inline-flex;
  align-items: center;
  background: aqua;
  box-shadow: 0 3px 2px 0 rgba(0, 0, 0, 0.1);
  border-radius: 5px;
  height: 45px;
  padding: 0 30px;
  color: #fff;
  font-family: Lato, Arial, sans-serif;
  text-transform: uppercase;
  text-decoration: none;
  transition: background .3s, transform .3s, box-shadow .3s;
  will-change: transform;
}
.button:hover {
  background: #ae4b67;
  box-shadow: 0 4px 17px rgba(0, 0, 0, 0.2);
  transform: translate3d(0, -2px, 0);
  cursor: pointer;
}
.button:active {
  box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.1);
  transform: translate3d(0, 1px, 0);
}

.pulse {
  position: relative;
}
.pulse:before, .pulse:after {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: rgba(255, 255, 255, 0.4);
  border-radius: 50%;
  width: 20px;
  height: 20px;
  opacity: 0;
  margin: auto;
}
.pulse:before {
  animation: pulse 1.5s infinite linear;
}
.pulse:after {
  animation: pulse 2s .4s infinite linear;
}
.pulse:hover:before, .pulse:hover:after {
  display: none;
}
```

> You will probably recognize everything except the actual styles to be applied and @keyframes. These would be easier learnt from https://w3schools.com. Now if you run the markup, the buttons would look pretty neat and the image would've resized itself. Time to give yourself a hooray!

Reviewer 3: link the keyframes tutorial on w3school

    console.log(firstNum + secondNum)
}
```
And here is a function without a parameter. It simply prints out 'hello world'.
```js
function sayHello(){
    console.log('Hello World!')
}
```
These can now be called from our HTML file as addNumbers(someNumber, someNumber) and sayHello(). Now that you have a basic understanding of how functions work, head over into your script.js file and add this in.
```js
var map = document.getElementById('map');
function changeImage(location){
    if (location == 'ocean'){
        map.src = 'https://cloud-9p8jb8vst.vercel.app/orgocean.png';
    }
    else if (location == 'residential'){
        map.src = 'https://cloud-gga696gyn.vercel.app/orgresidential.png';
    }
    
    else if (location == 'roadway'){
        map.src = 'https://cloud-bzeicw4mm.vercel.app/orgroad.png';
    }
    
    else if (location == 'vendors'){
        map.src = "https://cloud-9ldvvvhb4.vercel.app/orgvendors.png";
    }   
    
    else if (location == 'beach'){
        map.src = 'https://cloud-3bvmpy2n1.vercel.app/orgbeach.png';
    }
    else if (location == 'normal'){
        map.src = 'https://cloud-coh6o3mhy.vercel.app/orgbg.png';
    }
}
```

> This snippet declares a variable named map, which corresponds to the map element in our html file. Then it creates a function named changeImage() with the parameter 'location'. When we call this function from our html file with specific values like 'ocean' or 'residential', (These values passed into parameters are called arguments.) they call on the actual function from the JS file. Inside the function, is an IF -ELSE IF statement. It does exactly what it means. If the location argument is 'beach', set the map's src attribute to the beach highlighted image, and so on. The syntax is

Reviewer 3: ```suggestion
This snippet declares a variable named map, which corresponds to the map element in our html file. Then it creates a function named changeImage() with the parameter 'location'. When we call this function from our html file with specific values like 'ocean' or 'residential', (These values passed into parameters are called arguments.) they call on the actual function from the JS file. 

Inside the function, is an if else-if statement. It does exactly what it means. If the location argument is 'beach', set the map's src attribute to the beach highlighted image, and so on. The syntax is
```

```js
if (condition) {
    Do something;
}
else if (another condition) {
    Do something else;
}
else {
    Do something totally different;
}
```
There are three basic If statements in Javascript. The normal If statement, the If - Else statement, and the If - Else If - Else statement. Not sure what that means? Let's go through some examples.
```js
var name = 'Shane';
if (name == 'Shane'){
  console.log('Hello Shane')
}
```

> This snippet shows a basic If statement. It first declares a variable called name which has the value of shane. Then the If statement proceeds to check out if the name is Shane, and if yes, it logs out 'Hello Shane'. Note that nothing happens if the name is not set to Shane. To try this for yourself, create a seperate repl with the language set to Javascript only. Paste that snippet and tweak the values. Also, it's time to talk about the equal to symbols. Why only one equal to at the first and two equal to-s at the bottom? This is because one equal to set's the value of the left hand side to the right hand side and two equal to-s compare both the values. Here, we have used an equal to and set the value of name to Shane. Then we have used two equal to-s to check **if** the name and Shane are both the same.

Reviewer 3: ```suggestion
This snippet shows a basic if statement. It first declares a variable called name which has the value of shane. Then the if statement proceeds to check out if the name is Shane, and if yes, it logs out 'Hello Shane'. Note that nothing happens if the name is not set to Shane. To try this for yourself, create a seperate repl with the language set to Javascript only. Paste that snippet and tweak the values. Also, it's time to talk about the equal to symbols. Why only one equal sign at the first and two equal signs at the bottom? This is because one equal sign set's the value of the left hand side to the right hand side and two equal signs compare both the values. Here, we have used an equal sign and set the value of name to Shane. Then we have used two equal signs to check **if** the name and Shane are both the same.
```

Now, time for the next snippet. This one shows a basic If-Else statement.
```js
var pass = prompt('Enter your passcode: ');
if (pass == 'abcd'){
  console.log('Password Accepted');
}
else {
  console.log('Access Denied');
}
```
Now you probably wouldn't be knowing about prompt. Prompt is a function that allows us to take in user input. If you run this code in your Javascript-only repl, you'll get what I mean. Now let's dive into the code. It creates a variable named pass which gets user data. It then checks **if** the pass is 'abcd'. If the pass is abcd, it logs out that the password is accepted. **Else**, it logs out that access is denied. A quick word about prompt, prompt will appear different when you're using it with html. Here's a look.

![prompt in html/js](https://cloud-i7mb1y4jx.vercel.app/1243.jpg)

> Finally, the last If statement. This one checks **if** a condition is true, **else if** that condition is not true, it does something different, **else** it'll do something different. Here's a quick snippet. Be sure to run all these snippets so that you can see it live in-action.

Reviewer 3: make 'If' everywhere to lowercase.
